Decouple cmd and ApplyOptions Complete()

phanimarupaka opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**: This PR is to cleanup the way cobra cmd is used in ApplyOptions. The aim is to make Complete() method in ApplyOptions to not take cobra cmd as an input argument. By this change, we can decouple Complete() method in ApplyOptions from cobra command so that the ApplyOptions api is more reusable.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

phanimarupaka committed:
Decouple cmd and applyoptions complete

phanimarupaka commented:
@seans3 @mortent FYI

seans3 commented:
/ok-to-test

seans3 commented:
/assign

seans3 commented:
/retest

mortent commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -205,14 +208,12 @@ func NewCmdApply(baseName string, f cmdutil.Factory, ioStreams genericclioptions
-	if err != nil {
-		return err
+func (o *ApplyOptions) Complete(f cmdutil.Factory, flags cmdutil.AdditionalFlags) error {
+	if flags.DryRunStrategyResult.Err != nil {
```
It seems unnecessary to pass in the error here. We should just handle the error before calling the `Complete` function.

phanimarupaka replied:
Done

mortent commented on staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:
```diff
@@ -56,6 +56,17 @@ type debugError interface {
 	DebugError() (msg string, args []interface{})
 }
 
+type AdditionalFlags struct {
```
The word `Additional` here feels a bit vague. Maybe something like `ApplyFlags` would be better?

phanimarupaka replied:
Done

phanimarupaka commented:
/retest

phanimarupaka commented:
/retest

phanimarupaka commented:
/retest

soltysh commented:
/assign

brianpursley commented:
Looks like a good refactor.  

I do have a few questions though, and maybe this isn't something that has to be addressed in this PR, but wanted to bring them up:

1. Just so I fully understand, Is the intention that someone would be able to perform an apply without having a cobra.Command, by calling `ApplyOptions.Complete` and `ApplyOptions.Run` directly?

2. And if so, should validateArgs and validatePruneAll be moved out of `NewApplyOptions()` into  `Complete()` so that they too can be reused? Or do those validations only apply in a cobra command use case?

3. Finally, unit tests... many of the commands, including apply, test the entire `cobra.Command.Run()` but it sounds like that may be too high of a level to test if it is expected that people will be interacting with ApplyOptions directly.  Maybe that is beyond the scope of this change, but thought I would bring it up because it is at least somewhat related.

Overall, I like it though, just wondering if the refactor needs to go a little bit further to make sure all logic is moved out of `NewApplyOptions()` and include some additional unit tests possibly, for the non cobra command usage scenario.

phanimarupaka commented:
@brianpursley 

1. Yes. This helps decoupling Complete() from cobra which can be used by other applications as well.
2. validateArgs and validatePruneAll can also be called separately by other applications if they want to. For example, non-cli applications need not call validateArgs. Not sure if we need to move them into Complete()
3. The behavior of the application doesn't change with this change. This is just a level of abstraction and hence the existing tests would cover its correctness. I added test for the additional method I added.

brianpursley commented:
@phanimarupaka  OK thanks. 

So I guess for this command, `validateArgs()` is only checking command line arguments so it is OK for that to be in `NewCmdApply()`.

But I don't understand about `validatePruneAll()` because it looks like it validates the options, not the command line arguments:
```
cmdutil.CheckErr(validatePruneAll(o.Prune, o.All, o.Selector))
```

If I am writing a third party application that uses Apply and set `o.Prune` and `o.Selector` (an invalid combination), this validation will not be performed because `validatePruneAll()` is a private method only used inside `NewCmdApply()`.

phanimarupaka commented:
/retest

phanimarupaka commented:
@brianpursley Agree. Updated PR.

phanimarupaka commented:
/retest

brianpursley commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -172,9 +172,11 @@ func NewCmdApply(baseName string, f cmdutil.Factory, ioStreams genericclioptions
 		Example:               applyExample,
 		Run: func(cmd *cobra.Command, args []string) {
-			cmdutil.CheckErr(o.Complete(f, cmd))
+			o.RecordFlags.Complete(cmd)
```
Do you need to CheckErr this?
```
cmdutil.CheckErr(o.RecordFlags.Complete(cmd))
```

I see in the implementation that it always returns nil, but someone could change it in the future to actually return an error.

phanimarupaka commented:
/retest

brianpursley commented:
I was trying to think through how a non-cobra caller would use this, and it seems like there will be a lot of code required.  It **seems** like I will need to do all of these things:

* Create IOStreams (In, Out, ErrOut).
* Call NewApplyOptions.
* Create an instance of cmdutil.Factory (how do I create this without a cobra command?)
* Create an instance of cmdutil.ApplyFlags
* Call applyOptions.Complete(...) and check error
* Call applyOptions.Run() and check error

Is this correct?  Or maybe I don't see the simpler way that a non-cobra caller will use this.

It might be helpful to see an example of how a caller might use this after the refactor, without being in the context of a cobra command.  It doesn't need to be part of the commit, but maybe just in the comment or a gist would be nice.

I'm not trying to add more work, just trying to understand, so forgive me if I'm missing something.

seans3 commented:
/retest

phanimarupaka commented:
/retest

phanimarupaka commented:
/retest

phanimarupaka commented:
/retest

seans3 commented:
/retest

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:
```diff
@@ -172,9 +172,12 @@ func NewCmdApply(baseName string, f cmdutil.Factory, ioStreams genericclioptions
+			af, err := cmdutil.GetApplyFlags(cmd)
+			af.FieldManager = GetApplyFieldManagerFlag(cmd, o.ServerSideApply)
+			cmdutil.CheckErr(err)
+			cmdutil.CheckErr(o.Complete(f, af))
```
This is wrong, ideally this should have only 3 invocations:
1. Complete
2. Validate
3. Run.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go:
```diff
@@ -56,6 +56,14 @@ type debugError interface {
 	DebugError() (msg string, args []interface{})
 }
 
+type ApplyFlags struct {
```
This is wrong place for `ApplyFlags`, they should be placed in main apply.go. Compare WaitFlags: https://github.com/kubernetes/kubernetes/blob/d2fd467123aac0060d5cb3fd1a509adabc4ea190/staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go#L76
